### ISSUE TITLE
Using 4.23.0, my app fails to start because of invalid checksums

### ISSUE THREAD

**Issue by jbisotti** [liquibase/liquibase#4457]
### Search first

- [X] I searched and no similar issues were found

### Description

Our application (Java/Spring/MySQL) is currently using 4.20.0 and works fine. When I upgrade to 4.23.0, it fails to start due to Liquibase failing due to invalid checksums.

```
	Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.ValidationFailedException: Validation Failed:
     1 changesets check sum
          org/foo/db/phase5/foo-bar-2017-09-14.xml::DECREASE_FP_FIELD_FOR_PDP::jdoe was: 8:83871094f06fbeba7a3755bb6e7980c0 but is now: 8:7afe9a1d89ce253fbca2d513dcdc2834
```

### Steps To Reproduce

The offending changeset (names changed to protect the innocent):

```xml
    <changeSet author="jdoe" id="DECREASE_FP_FIELD_FOR_PDP" context="bootstrap,demo,unit">
        <comment>Words words words. Please review if your institution would like this change. See JIRA-123 for details.</comment>

        <sql dbms="mysql">
            update FP_DV_PAYEE_DTL_T A join FP_DV_PAYEE_DTL_T B ON A.FDOC_NBR = B.FDOC_NBR set A.DV_RMT_PRSN_NM = SUBSTR(B.DV_RMT_PRSN_NM,1,75);
        </sql>

        <sql dbms="oracle">
            update FP_DV_PAYEE_DTL_T A set DV_RMT_PRSN_NM = (select SUBSTR(DV_RMT_PRSN_NM,1,75) from FP_DV_PAYEE_DTL_T B where A.FDOC_NBR = B.FDOC_NBR);
        </sql>

        <modifyDataType columnName="DV_RMT_PRSN_NM" newDataType="VARCHAR(75)" tableName="FP_DV_PAYEE_DTL_T"/>
    </changeSet>
```

### Expected/Desired Behavior

My app would startup like it does with 4.20.0.

### Liquibase Version

4.23.0

### Database Vendor & Version

MySQL 5.7

### Liquibase Integration

Spring (not Boot)

### OS and/or Infrastructure Type/Provider

macOS

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by kevin-atx** [comment]
@filipelautert  - passing this along to you since it's likely related to the 4.23.0 checksum change

**Comment by filipelautert** [comment]
Hello @jbisotti - I saw you're using 2 dbms tags within the same changeset so your problem is related to the 4.23.0 bug that will be fixed by PR https://github.com/liquibase/liquibase/pull/4453 .
Meanwhile there are 3 workarounds to upgrade:

1. add tag` <validCheckSum>1:any</validCheckSum> `to your changeset, upgrade to 4.23.0 then remove it.
2. run `clearChecksums ` command and then upgrade.
3. Update db changelog table using the new checksum: `update databasechangelog set md5sum = '8:7afe9a1d89ce253fbca2d513dcdc2834' where md5sum = '8:83871094f06fbeba7a3755bb6e7980c0' .`

**Comment by filipelautert** [comment]
@jbisotti I took a better look at PR https://github.com/liquibase/liquibase/pull/4453 and it wo'nt fix your checksum problem - it only fixes a bug with your setup in 4.23.0.  So I would suggest for you to just upgrade after that PR is merged as if you use changesets in the same way you set up your example one (2 different dbms on the same changeset) they may not be executed.

**Comment by jacokok** [comment]
We have same problem but it gets new md5sum every time we run update-testing-rollback.

It only happens on changesets where we are using parameters in liquibase.properties file.

Will this be resolved by same fix or should I create new issue?

**Comment by filipelautert** [comment]
Hello @jacokok ! Does it solve your problems if you try any of the steps above? Just trying to understand what would be the correct pr for it.

**Comment by jacokok** [comment]
It fixes it until I run `liquibase update-testing-rollback` then it gives me the old md5sum again.

Currently we need to follow the steps above every time we run `liquibase update` after `liquibase update-testing-rollback` was executed.

**Comment by jacokok** [comment]
This happens on change types where we are using properties like `${databaseName}` in this example.
After rollback it will have new md5sum. When you clear md5sum it runs fine until the next rollback. 

```yaml
  - changeSet:
      id: memoryFile
      author: Redacted
      comment: Add File To Use In Memory Types
      runInTransaction: false
      changes:
        - sql:
            sql: |
              IF NOT EXISTS(SELECT TOP 1 1 FROM sys.database_files WHERE name = '${databaseName}_Memory')
              BEGIN
                ALTER DATABASE CURRENT ADD FILE (name='${databaseName}_Memory', filename='${memoryFile}') TO FILEGROUP FirstMap_Memory                 
              END
      rollback: empty
```

```properties
parameter.databaseName=DatabaseName
```

For now I added the below to solve issue.

```yaml
    validCheckSum:
      - "1:any"
```

**Comment by jacokok** [comment]
@filipelautert even with v9 checksums I am still getting errors after running update-testing-rollback.

`was: 9:0c23b871a754652fd0769fbb2b64d140 but is now: 9:e815bcec1cb06046ed36e74dbfab7b09`

**Comment by jbisotti** [comment]
> @jbisotti I took a better look at PR https://github.com/liquibase/liquibase/pull/4453 and it wo'nt fix your checksum problem - it only fixes a bug with your setup in 4.23.0. So I would suggest for you to just upgrade after that PR is merged as if you use changesets in the same way you set up your example one (2 different dbms on the same changeset) they may not be executed.

Sorry, I do not understand what you are trying to tell me here.

This is not a new changeSet; it is not a contrived example; it is a production changeSet that has been part of our software since 2017. We update Liquibase frequently; no version up through 4.20.0 has had any issues with this changeSet.

I want to be able to continue easily upgrading to the latest version of Liquibase and it "just work". We don't want to
1. Put workarounds in years old changeSets
2. Clear out years worth of checksums
   * We only use Liquibase as part of our app startup; no separate/external usage.
   * Our maintenance window is tight as it is already.
   * Due to SOC, we cannot change a prod DB outside of the app.
3. Manually muck with the `datbasechangelog` table
   * This feels gross. 
   * Due to SOC, we cannot change a prod DB outside of the app.

Are you telling me upgrading beyond 4.20.0 will no longer "just work" for us?

**Comment by filipelautert** [comment]
@jbisotti |it was really confusing. So:
* 4.21.1 must be fine for you now
* Don't upgrade to 4.23.0 because of this bug 
* Once PR https://github.com/liquibase/liquibase/pull/4453 is merged you can upgrade (probably 4.23.1 or 4.24.0) .

**Comment by jacokok** [comment]
Perfect thanks for the info!

**Comment by Alf-Melmac** [comment]
This issue was not fixed in version 4.32.1 (where the mentioned pr https://github.com/liquibase/liquibase/pull/4453 was merged into). This can be reproduced if you create a changeset where the dbms of the sql statement doesn't match your current database.
For example in an environment with postgres
```xml
<changeSet id="broken.a" author="test">
	<sql dbms="!postgresql">
		-- Some SQL
	</sql>
</changeSet>

<changeSet id="broken.b" author="test">
	<sql dbms="mssql">
		-- Some SQL
	</sql>
</changeSet>
```
Use liquibase version 4.21.1 and execute `mvn liquibase:update`. Afterwards update to the current version 4.23.2. I've created a minimal reproducible example: https://github.com/Alf-Melmac/liquibase-issue-checksums with some more examples in the [changelog.xml](https://github.com/Alf-Melmac/liquibase-issue-checksums/blob/main/src/main/resources/changelog.xml)
